### PR TITLE
Add celotool local-testnet command as an easy way to create local testnets

### PR DESCRIPTION
**Pull request by nategraf**
### Description

This PR adds the `celotool local-testnet` command which can be used to quickly and easily create a local testnet.

### Usage

Running `yarn celotool local-testnet --help` is the fastest way to get started. Here is the description the user will see:

```
Running this command will create a number of geth nodes, connect them together to form a network,
and run smart contract migrations to initialize the core protocols. When this is complete, it will
open a NodeJS console with some preloaded objects to facilitate interactions with the test network.
Exiting this console will kill all running geth instances and exit.

Examples:
* local-testnet
* local-testnet --local-geth ~/code/celo-blockchain
* local-testnet --validators 5 --proxies 3 --bootnode
* local-testnet --tx-nodes 2 --light-clients 3
* local-testnet --migrate-to 19 --migration-override '{ "lockedGold": { "unlockingPeriod": 30 } }'
* local-testnet --no-migrate --genesis-override '{ "blockTime": 3, "epoch": 50 }'

Network makeup is configured the --validators, --tx-nodes, --light-clients, and --lightest-client
flags. These flags will add the corresponding nodes to the network.

A NodeJS REPL is provided to conveniently interact with the created network. A number of global
variables are defined with useful values including {
  Web3 (Imported 'web3' module)
  Admin (Imported 'web3-eth-admin' module)
  testnet: GethRunConfig (Configuration values for the tesnet)
  [nodeType][index] (e.g. validator0, txNode2): {
    web3: Web3 (A web3 object connected to the node over RPC)
    admin: Admin (An Admin object connected to the node over RPC)
    config: GethInstanceConfig (Configuration values for the node)
    kill(signal?: string): (Send a signal, default SIGTERM, to the node. e.g. SIGINT, SIGSTOP)
  }
}

Tip: Export NODE_OPTIONS="--experimental-repl-await" in your terminal to natively use await.

When the network is created without a bootnode, all nodes will be connected as follows:
* Proxy nodes are connected to their validators, other proxies and unproxied validators.
* Unproxied validator nodes are connected to all other proxies and unproxied validators.
* Transaction nodes are connected to proxies and unproxied validators and other transaction nodes.
* Light clients are connected to all transaction nodes

If the network is started with the --bootnode flag, a bootnode will be created and all node will be
connected to it, rather than each other directly.

By default, the celo-blockchain repository will be cloned to a temporary location and built from
master to produce the geth binary to run for each node. The --branch flag can be used to control
which branch is built in the cloned repository. Alternatively, a existing repository can be used
by specifying the --local-geth flag as the path to that repository root.
```

### Other changes

* Modifications to the Geth config types for end-to-end testing
* Improve utilities for sending signals to and killing processes
* Refactor the interconnection of nodes in e2e tests for better flexibility

### Tested

I have been using this branch in my personal workflows for the past year.

### Backwards compatibility

100%

### CODE REVIEW

File: packages/celotool/src/lib/geth.ts [hunk 1]
```diff
@@ -1145,39 +1159,67 @@ export function spawnWithLog(cmd: string, args: string[], logsFilepath: string,
...
+  await connect(left, right)
+  await connect(right, left)
 }
 
 // Add validator 0 as a peer of each other validator.
 export async function connectValidatorPeers(instances: GethInstanceConfig[]) {
-  await connectPeers(
-    instances.filter(({ wsport, rpcport, validating, isProxy, isProxied }) => ((validating && !isProxied) || isProxy)  && (wsport || rpcport))
+  const validators = instances.filter(
+    (node) => ((node.validating && !node.isProxied) || node.isProxy)
```
Comment (yorhodes) [inline packages/celotool/src/lib/geth.ts]: nice

File: packages/celotool/src/cmds/local_testnet.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+  extend(session.context, {
+    Web3,
+    Admin,
+    testnet: config,
+    ...config.instances.reduce(
+      (o, instance) => ({
+        ...o,
+        [formatName(instance.name)]: {
+          web3: new Web3(getRpcUrl(instance)),
```
Comment (yorhodes) [inline packages/celotool/src/cmds/local_testnet.ts:174]: this is amazing
wonder if it's worthwhile in future to add a contractkit here too

File: packages/celotool/src/e2e-tests/governance_tests.ts [hunk 3]
```diff
@@ -9,7 +9,12 @@ import BigNumber from 'bignumber.js'
 import { assert } from 'chai'
 import path from 'path'
 import Web3 from 'web3'
-import { connectPeers, connectValidatorPeers, importGenesis, initAndStartGeth } from '../lib/geth'
+import {
+  connectBipartiteClique,
+  connectPeers,
+  /*connectValidatorPeers,*/ importGenesis,
```
Comment (yorhodes) [inline packages/celotool/src/e2e-tests/governance_tests.ts]: remove?

File: packages/celotool/src/e2e-tests/governance_tests.ts [hunk 4]
```diff
@@ -967,6 +971,7 @@ describe('governance tests', () => {
       console.log('wait for end of epoch')
       // Wait for an epoch transition to ensure everyone is connected to one another.
       await waitForEpochTransition(web3, epoch)
+      await sleep(5.5)
```
Comment (yorhodes) [inline packages/celotool/src/e2e-tests/governance_tests.ts]: is this necessary?

File: packages/celotool/src/e2e-tests/utils.ts [hunk 5]
```diff
@@ -114,36 +114,52 @@ export function assertAlmostEqual(
...
+export async function killInstance(instance: GethInstanceConfig) {
+  if (instance.pid) {
+    await shutdownOrKill(instance.pid)
+  }
+}
+
+export async function shutdownOrKill(identifier: string | number) {
+  await signalProcess(identifier, 'INT')
 
+  // Poll for remaining processes for up to ~30s with exponetial backoff.
```
Comment (yorhodes) [inline packages/celotool/src/e2e-tests/utils.ts]:
```suggestion
  // Poll for remaining processes for up to ~30s with exponential backoff.
```

**Review by yorhodes — APPROVED** [review]
not sure I understand the network topology of bipartite clique with sources and targets
some docs or a diagram would be epic!

File: packages/celotool/src/e2e-tests/governance_tests.ts (hunk 3 above)
Comment (nategraf) [reply packages/celotool/src/e2e-tests/governance_tests.ts]: Done

File: packages/celotool/src/e2e-tests/governance_tests.ts (hunk 4 above)
Comment (nategraf) [reply packages/celotool/src/e2e-tests/governance_tests.ts]: Removed

File: README-dev.md [hunk 6]
```diff
@@ -1,5 +1,16 @@
 # README GUIDE FOR CELO DEVELOPERS
 
+## How to create a local testnet
+
+Often when developing, it is useful to create a test network localy using the full celo-blockchain binary to go beyond what can be done with other options such as [Ganache](https://www.trufflesuite.com/ganache)
+
+The `celotool` package provides two options for creating local testnets. One is `celotool local-testnet` and the other is `celotool geth start`.
+
```
Comment (nategraf) [inline README-dev.md]:
```suggestion
```

File: README-dev.md [hunk 7]
```diff
@@ -1,5 +1,16 @@
 # README GUIDE FOR CELO DEVELOPERS
 
+## How to create a local testnet
```
Comment (nategraf) [inline README-dev.md]:
```suggestion
## How to run a local testnet
```

File: packages/celotool/src/cmds/local_testnet.ts [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+Network makeup is configured the --validators, --tx-nodes, --light-clients, and --lightest-client
+flags. These flags will add the corresponding nodes to the network.
+
+A NodeJS REPL is provided to conveniently interact with the created network. A number of global
+variables are defined with useful values including {
+  Web3 (Imported 'web3' module)
+  Admin (Imported 'web3-eth-admin' module)
+  testnet: GethRunConfig (Configuration values for the tesnet)
+  [nodeType][index] (e.g. validator0, txNode2): {
+    web3: Web3 (A web3 object connected to the node over RPC)
```
Comment (nategraf) [inline packages/celotool/src/cmds/local_testnet.ts:38]:
```suggestion
    web3: Web3 (A web3 object connected to the node over RPC)
    kit: ContractKit (A contractkit object connected to the node over RPC)
```

File: packages/celotool/src/cmds/local_testnet.ts [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+  }
+}
+
+Tip: Export NODE_OPTIONS="--experimental-repl-await" in your terminal to natively use await.
+
+When the network is created without a bootnode, all nodes will be connected as follows:
+* Proxy nodes are connected to their validators, other proxies and unproxied validators.
+* Unproxied validator nodes are connected to all other proxies and unproxied validators.
+* Transaction nodes are connected to proxies and unproxied validators and other transaction nodes.
+* Light clients are connected to all transaction nodes
```
Comment (nategraf) [inline packages/celotool/src/cmds/local_testnet.ts]:
```suggestion
* Light clients are connected to all transaction nodes.
```

File: packages/celotool/src/cmds/local_testnet.ts [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+If the network is started with the --bootnode flag, a bootnode will be created and all node will be
```
Comment (nategraf) [inline packages/celotool/src/cmds/local_testnet.ts]:
```suggestion
If the network is started with the --bootnode flag, a bootnode will be created and all nodes will be
```

File: packages/celotool/src/cmds/local_testnet.ts (hunk 2 above)
Comment (nategraf) [reply packages/celotool/src/cmds/local_testnet.ts:174]: I added a contractkit instance 👌

**Pull request merged**